doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+impossible to implement for a user-defined DType to register universal
+function loops.
+
+This NEP proposes a new ``UFuncImpl`` struct, with an extensible C-API
+to register more feature rich loops, thus allowing these use-cases.
+It also proposes to address certain deficiencies, for example with respect
+to error reporting, by extending the signature of the core inner-loop
+functions.
+Finally, one of the most important steps in the ufunc call is finding the
+correct function for a specific set of input dtypes.  We propose to use
+multiple-dispatching in the future.
+
+.. note::
+
+    At this time this NEP is in a preliminary state. Both internal and
+    external API may be adapted based on user input or implementation needs.
+    The general design principles and choices, while provisional, should not
+    be expected to change dramatically.
+    For example, the *exact* additional information passed into ``UFuncImpl``
+    is likely to evolve.
+
+
+Detailed Description
+--------------------
+
+NEP 41 and NEP 42 layed out the general proposed structure into DType
+classes (similar to the current type numbers) and dtype instances.
+Universal functions have a fairly fixed signature of multiple input arrays
+and certain additional features. But depending on the input DTypes, they
+have to perform a different operation.  In general these operations are
+registered by users, e.g. implementers of a user-defined DTypes.
+
+A UFunc call consists of into multiple steps:
+
+1. Resolution of ``__array_ufunc__`` for container types, such as a Dask
+   array handling the full process, rather than NumPy.
+   This step is performed first, and unaffected by this NEP.
+
+2. *Promotion and Dispatching*
+
+   * Given the DTypes of all inputs we need to find the correct implementation
+     for the ufuncs functionality. E.g. an implementation for ``float64``
+     or ``int64``, but also a user-defined DType.
+
+   * When no exact implementation exists, *promotion* has to be performed.
+     For example, adding ``float32`` and ``float64`` is implemented by
+     first casting the ``float32`` to ``float64``.
+
+3. *DType Adaptation:*
+
+   * The step has to perform no special work for non-parametric dtypes.
+   * For example, if a loop adds two strings, it is necessary to define the
+     correct output (and possibly input) dtypes.  ``S5 + S4 -> S9``, while
+     an ``upper`` function has the signature ``S5 -> S5``.
+
+4. Preparing the actual iteration. This step is largely handled by ``NpyIter`` (the iterator).
+
+   * Allocate all ouputs and potentially copies (or buffers).
+   * Finds the best iteration order, which includes information such as
+     a broadcasted stride always being 0.
+
+5. *Loop-specific setup* may include for example:
+
+   * Clearing of floating point exception flags (if necessary),
+   * Possibly allocating temporary working space,
+   * Setting (and potentially finding) the inner-loop function.  Finding
```

Comment:
True, although I think this can be part of the setup?  This is about specializations, such as contiguous loops (SIMD capable strides), or reduction loops, which can be implemented faster.  I guess the grouping here is trying to find the most practical grouping for users who want the most control possible, while not asking them to take charge of setting up the outer iteration as well.

Anyway, thanks for this and all the feedback, will try to iterate/think on it!

Reply:
Ah, with that goal it does make sense. I think then all that's needed is to give exactly that example!